Clustering

freeekanayaka opened the pull request:
Add clustering support to LXD.

This is a work in progress, and the goal of this PR is mainly to see what changes are needed to make tests pass on Jenkins.

stgraber commented:
@freeekanayaka Jenkins should now be able to run the testsuite for clustering, but your branch seems to be broken in the `lxc list` tests right now.

freeekanayaka commented:
@stgraber the unit test failure is now fixed.

However there's another failure when trying to start the lxd daemon:

error: failed to open cluster database: gRPC could not send open request: gRPC OPEN response error: rpc error: code = Unknown desc = failed to handle OPEN request: could not open 

I'm not able to reproduce it here. Could you please try to run the integration test suite on your end and see if it happens?

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
@stgraber I temporarily disabled the clustering-specific integration tests that make use of lxc-execute/nsenter & co. Now tests are basically green. As mentioned before, I guess some time next week I'll need some assistance to understand how to make the Jenkins environment happy about those disabled integration tests.

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

stgraber commented:
Random comment while playing with it:
 - Can we rename "node_name" to "server_hostname" in the environment dict (and always fill it either with the hostname of the system or the chosen node name in the cluster)
 - Can we similarly rename "clustered" to "server_clustered"?

That should make my OCD slightly happier when reading the current environment :)

stgraber commented:
Also can we make it so that we don't prompt users for node-specific keys when no host has that key set (example is when "source" isn't set on a storage pool or "bridge.external_interfaces" isn't set on a network)?

stgraber commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

stgraber commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
@stgraber your comments should be all addressed. However, note that when you create a storage pool on the bootstrap node via lxd init, in several cases the "source" config key gets populated even if you don't really specify it explicitly (e.g. for the dir backend, or for btrfs/zfs loopback), so at the moment in those cases you'll still be asked for that when joining the cluster interactively.

freeekanayaka commented:
@stgraber I also think that tests are (more?) stable now. The only problem that I see is a cgo-related crash when running i386/go-tip. That's a bit nasty, and will require some work on the go-sqlite3/dqlite side. It shouldn't really be a blocker tho.

freeekanayaka commented:
Hm, actually I just saw the cgo-related crash also on 1.9/i386, that's the first time I believe. However, I never saw it on amd64. As said, it will require some work, lemme know if you consider this blocker.

stgraber commented:
@freeekanayaka would be nice if you could spend some time looking into that cgo crash before we merge this on Thursday. As it seems to hit pretty often, it'd certainly affect the team's ability to get good test results for our branches...

stgraber commented:
jenkins: test this please

freeekanayaka commented:
@stgraber I'll try, it just requires some internal rework since Go is pretty picky these days about passing pointers/data back and forth from C.

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

stgraber commented on doc/api-extensions.md:
```diff
@@ -415,3 +415,31 @@ same and between storage pools.
+This includes the following new endpoints:
+
+* `GET /1.0/cluster`
+* `DELETE /1.0/cluster` (see [RESTful API](rest-api.md) for details)
```
Should be "UPDATE /1.0/cluster" and also be used to initialize the cluster or join an existing cluster (see other comments).

freeekanayaka replied:
Added.

stgraber commented on doc/api-extensions.md:
```diff
@@ -415,3 +415,31 @@ same and between storage pools.
+* `POST /1.0/cluster/nodes` (see [RESTful API](rest-api.md) for details)
+
+* `GET /1.0/cluster/nodes/<name>` (see [RESTful API](rest-api.md) for details)
+* `DELETE /1.0/cluster/nodes/<name>` (see [RESTful API](rest-api.md) for details)
```
Rename "nodes" to "members" in all of the above URLs (sorry for the messing this up in the initial design...)

freeekanayaka replied:
Done.

stgraber commented on doc/rest-api.md:
```diff
@@ -217,6 +217,9 @@ won't work and PUT needs to be used instead.
      * `/1.0/resources`
+     * `/1.0/cluster`
+       * `/1.0/cluster/nodes`
+         * `/1.0/cluster/nodes/<name>`
```
s/nodes/members/

freeekanayaka replied:
Done.

stgraber commented on doc/rest-api.md:
```diff
@@ -2424,3 +2427,236 @@ Input (none at present):
+    }
+
+## `/1.0/cluster`
+### GET (optional `?password=<trust-password>`)
```
password will most likely be dropped (see other comments)

freeekanayaka replied:
Do you mean we should just drop it from the doc? (although the current implementation does that)

stgraber replied:
We should drop that password field from the API too. Our workflow should not involve passing a password as an API argument.

freeekanayaka replied:
The password query parameter was dropped, as per the new API layout/semantics.

stgraber commented on doc/rest-api.md:
```diff
@@ -2424,3 +2427,236 @@ Input (none at present):
+        "operation": "",
+        "error_code": 0,
+        "error": "",
+        "metadata": {
```
So I think I'd expect something like:

```json
            "enabled": true,
            "status": "RUNNING",
            "current_server": "lxd01",
            "member_config": [
                {
                     "name": "storage-pools.default.config.source",
                     "description": "\"source\" property for storage pool \"default\""
                },
                {
                     "name": "networks.lxdbr0.config.bridge.external_interfaces",
                     "description": "\"bridge.external_interfaces\" property for network \"lxdbr0\""
                }
           ]
```

stgraber replied:
I don't understand, the stuff listed above should be the only thing we actually need in the local database, everything else is in the global one. So when joining, ask for those things, put them in the local database and now when you merge the global and local bits for that storage or network, you've got everything you need.

stgraber replied:
As discussed earlier, lets keep this for a further update of the API where we add a "nice" way to join the cluster. Until then, we'll just be using GetNetworks, GetStoragePools on the cluster side and CreateNetwork and CreateStoragePool on the local side.

stgraber commented on doc/rest-api.md:
```diff
@@ -2424,3 +2427,236 @@ Input (none at present):
+    	}
+    }
+
+### DELETE
```
Turning it off should be PUT with enabled=false

freeekanayaka replied:
I've changed this to PUT, but I didn't use the enabled flag, since it's effectively redundant: if Name is not an empty string, then enable, otherwise disable.

stgraber commented on doc/rest-api.md:
```diff
@@ -2424,3 +2427,236 @@ Input (none at present):
+	} 
+
+### POST
+ * Description: bootstrap, join, or accept a node in the cluster
```
Should be limited to adding (accepting) a node. Bootstrap and join should go through "POST /1.0/cluster" instead.

freeekanayaka replied:
Fixed as per the new API layout.

stgraber commented on doc/rest-api.md:
```diff
@@ -2424,3 +2427,236 @@ Input (none at present):
+		}
+	}
+
+## `/1.0/cluster/nodes/<name>`
```
What's going on with this one? Looks like some bad copy/paste from the storage volume one.

freeekanayaka replied:
I'm not sure what you mean. I did take storage as template, but the info here is just documenting input/output formats for POST /1.0/cluster/members

stgraber commented on shared/api/storage_pool_volume.go:
```diff
@@ -31,6 +31,7 @@ type StorageVolume struct {
 	Name             string   `json:"name" yaml:"name"`
 	Type             string   `json:"type" yaml:"type"`
 	UsedBy           []string `json:"used_by" yaml:"used_by"`
+	Node             string   `json:"node" yaml:"node"`
```
Should be "Location" instead and should be under an "API extension" comment too.

freeekanayaka replied:
Done

stgraber commented on shared/api/storage_pool.go:
```diff
@@ -19,6 +19,10 @@ type StoragePool struct {
+
+	// API extension: clustering
+	State string   `json:"state" yaml:"state"`
+	Nodes []string `json:"nodes" yaml:"nodes"`
```
Should be Location instead of Nodes

freeekanayaka replied:
Just to double check: "Location" singular?

freeekanayaka replied:
Went for "Location" for now.

stgraber replied:
Locations would probably be best indeed given it's a list.

stgraber replied:
done

stgraber commented on shared/api/server.go:
```diff
@@ -16,6 +16,10 @@ type ServerEnvironment struct {
+
+	// API extension: clustering
+	ServerClustered bool   `json:"server_clustered" yaml:"server_clustered"`
+	ServerName      string `json:"server_name" yaml:"server_name"`
```
How about we keep those two under /1.0/cluster? (would require a GET /1.0/cluster to retrieve the "enabled" and "server_name" properties)

freeekanayaka replied:
It's used in a few places and would require some change. But sure, if you prefer.

stgraber replied:
Okay, fine to keep where it is, it'll be useful in bug reports as we always request the `lxc info` output.

stgraber commented on shared/api/network.go:
```diff
@@ -38,6 +38,10 @@ type Network struct {
+
+	// API extension: clustering
+	State string   `json:"state" yaml:"state"`
+	Nodes []string `json:"nodes" yaml:"nodes"`
```
Should be Location instead of Nodes

freeekanayaka replied:
So since it's the second time you spell it singular I guess you really mean it :). Done.

stgraber replied:
We seem to have been consistently using the plural form for lists, so Locations is actually better, sorry :)

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,58 @@
+type Cluster struct {
+	StoragePools []StoragePool `json:"storage_pools" yaml:"storage_pools"`
+	Networks     []Network     `json:"networks" yaml:"networks"`
+}
```
This should instead be more like:

```go
type ClusterMemberConfig struct {
    Name string
    Description string
}

type Cluster struct {
    Enabled bool
    Status string
    MemberConfig []ClusterLocalConfig
    CurrentServer string
}

// Very rough, will update once we sort out the right workflow there
type ClusterPut struct {
   Enabled bool
   MemberConfig map[string]string
   Cert string
   Key string
   RaftNodes ClusterRaftNodes
}
```

freeekanayaka replied:
As said above, ClusterMemberConfig might not be the best format for how this info is intended to be consumed.

stgraber commented on shared/api/network.go:
```diff
@@ -38,6 +38,10 @@ type Network struct {
 	Managed bool `json:"managed" yaml:"managed"`
+
+	// API extension: clustering
+	State string   `json:"state" yaml:"state"`
```
Should be Status (and "status" in API) for consistency

stgraber replied:
Containers have always had a status field as do operations. Both of those use Status and StatusCode for that.

freeekanayaka replied:
This was fixed in 73d0351a1ce6716830848edb68499ef23b61c41c.

stgraber commented on shared/api/storage_pool.go:
```diff
@@ -19,6 +19,10 @@ type StoragePool struct {
 	UsedBy []string `json:"used_by" yaml:"used_by"`
+
+	// API extension: clustering
+	State string   `json:"state" yaml:"state"`
```
Should be Status (and "status" in API) for consistency

freeekanayaka replied:
This was fixed in 73d0351a1ce6716830848edb68499ef23b61c41c.

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,58 @@
+// ClusterNodeAccepted represents the response of a request to join a cluster.
+//
+// API extension: cluster
+type ClusterNodeAccepted struct {
```
`ClusterMembersPostResponse`

freeekanayaka replied:
Done.

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,58 @@
+// RaftNode represents the a LXD node that is part of the dqlite raft cluster.
+//
+// API extension: cluster
+type RaftNode struct {
```
`ClusterRaftNodes`

stgraber replied:
I'm not crazy about having raft be mentioned in our public API, especially when there doesn't really appear to be much of anything raft specific there. Is the Address there the same as the URL for the member?

If so, how about we just add the ID to ClusterMember and have the join response include a full list of all members, would that get us what you need?

freeekanayaka replied:
Not really. This list must hold essentially two things 1) IP addresses of database nodes (which yes, is roughly the same as ClusterMember.URL) 2) their raft server ID (this ID is a raft internal thing, unrelated to our database ID for ClusterMember).

We could probably rename RaftNode to ClusterDatabaseMember or something like that, if you wish.

stgraber replied:
How about adding a DatabaseID to the ClusterMember struct which would contain that raft node id when Database is true?

Would that then let you use a []ClusterMember when joining? Having this information visible in the API would also make it easier to debug some logs if the raft node id shows up in there.

freeekanayaka replied:
This has been moved to an internal API, so it's not anymore public.

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,58 @@
+// NodePost represents the fields required to rename a LXD node.
+//
+// API extension: clustering
+type NodePost struct {
```
`ClusterMemberPost`

freeekanayaka replied:
Done.

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,58 @@
+// Node represents the a LXD node in the cluster.
+//
+// API extension: clustering
+type Node struct {
```
`ClusterMember`

freeekanayaka replied:
Done.

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,58 @@
+	Name     string `json:"name" yaml:"name"`
+	URL      string `json:"url" yaml:"url"`
+	Database bool   `json:"database" yaml:"database"`
+	State    string `json:"state" yaml:"state"`
```
`Status`

freeekanayaka replied:
See above.

freeekanayaka replied:
This was fixed in 73d0351a1ce6716830848edb68499ef23b61c41c.

freeekanayaka commented on Makefile:
```diff
@@ -105,6 +105,14 @@ update-pot:
+	    echo $(shell git log -1 --format=format:%H) > manifest.uuid && \
+	    ./configure && \
+	    make
+
```
Leftover, dropped.

freeekanayaka commented on client/events.go:
```diff
@@ -98,3 +98,13 @@ func (e *EventListener) Wait() error {
 }
+
+// Active returns true if this listener is still connected, false otherwise.
+func (e *EventListener) Active() bool {
```
Renamed.

freeekanayaka commented on client/interfaces.go:
```diff
@@ -45,9 +45,13 @@ type ContainerServer interface {
 	// Server functions
 	GetServer() (server *api.Server, ETag string, err error)
 	GetServerResources() (resources *api.Resources, err error)
+	GetServerHost() (host string, err error)
```
Yeah this is a bit unfortunate indeed. The problem is that GetConnectionInfo requires r.server to be populated, hence to perform the initial automatic GetServer request, and in some cases that is problematic, see the comment below.

stgraber replied:
Let's remove GetServerHost and replace with GetConnectionInfo, modifying it to skip the Addresses handling if r.server == nil

stgraber replied:
Done

freeekanayaka commented on client/connection.go:
```diff
@@ -45,6 +45,9 @@ type ConnectionArgs struct {
 	CookieJar http.CookieJar
+
+	// Skip automatic GetServer request upon connection
+	SkipGetServer bool
```
Yeah I know, but I don't have many other solutions since the "lxd shutdown" command uses lxd.ConnectLXDUnix to send a raw request to /internal/shutdown. Now, if that would involve a database query (like /api/1.0 does), you would not be able to shutdown the last node of a cluster (since it has lost leadership/quorum).

freeekanayaka replied:
I guess I could think of a few tricks to avoid exposing this in ConnectionArgs or interfaces.go, if that's desired. Or we could just add a comment that this is for internal LXD usage, ISTR there are other client interfaces marked as such.

stgraber replied:
Hmm so isn't there a bigger problem with the shutdown logic when clustered then?
Calling `lxd shutdown` will trigger a clean shutdown of all containers, which certainly involves database access.

stgraber replied:
For shutdown, lets do this:
 - Keep a list of running containers in memory, something like:
```go
type runningContainer struct {
    Name string
    StopPriority int
    HostShutdownTimeout int
}
```
 - Modify containersShutdown to use that struct
 - For each of those, we need to load them through go-lxc lxc.NewContainer
 - Shutdown and final Stop should then be handled through go-lxc in the clustered case, eliminating all DB calls
 - Modify internalContainerOnStop to know that we're shutting down and that the database isn't available, in such case, it's fine for it to be a no-op

freeekanayaka replied:
As agreed during the catch-up, this is not blocking and will be done after this initial PR is merged.

freeekanayaka commented on client/interfaces.go:
```diff
@@ -45,9 +45,13 @@ type ContainerServer interface {
 	HasExtension(extension string) (exists bool)
 	RequireAuthenticated(authenticated bool)
+	IsClustered() (clustered bool)
+	ClusterTargetNode(name string) (client ContainerServer)
```
Renamed it to UseTarget. If you prefer SwitchTarget, it should be pretty trivial.

freeekanayaka commented on client/interfaces.go:
```diff
@@ -45,9 +45,13 @@ type ContainerServer interface {
 	RequireAuthenticated(authenticated bool)
+	IsClustered() (clustered bool)
+	ClusterTargetNode(name string) (client ContainerServer)
+	ClusterNodeName() (name string)
```
Renamed.

freeekanayaka commented on client/lxd.go:
```diff
@@ -110,6 +113,13 @@ func (r *ProtocolLXD) RawWebsocket(path string) (*websocket.Conn, error) {
 	return r.websocket(path)
 }
 
+// RawOperation allows direct creation of LXD API operations.
```
Amended as you suggest.

freeekanayaka commented on doc/rest-api.md:
```diff
@@ -2424,3 +2427,236 @@ Input (none at present):
             }
         }
     }
+## `/1.0/storage-pools`
```
Probably some copy/paste or rebase leftover. Dropped.

freeekanayaka commented on client/interfaces.go:
```diff
@@ -164,9 +168,20 @@ type ContainerServer interface {
+	JoinCluster(targetAddress, targetPassword, targetCert, name string) (op *Operation, err error)
+	LeaveCluster(name string, force bool) (err error)
+	GetNodes() (nodes []api.Node, err error)
+	GetNode(name string) (node *api.Node, err error)
```
Done. For sake of getting this done, I'm returning an empty string right now, I guess we can implement properly it after the merge.

stgraber replied:
Yeah, should just need to be changed to an ETag response in the server.

stgraber replied:
done

freeekanayaka commented on client/interfaces.go:
```diff
@@ -164,9 +168,20 @@ type ContainerServer interface {
+	BootstrapCluster(name string) (op *Operation, err error)
+	AcceptNode(targetPassword, name, address string, schema, api int, pools []api.StoragePool, networks []api.Network) (info *api.ClusterNodeAccepted, err error)
+	JoinCluster(targetAddress, targetPassword, targetCert, name string) (op *Operation, err error)
+	LeaveCluster(name string, force bool) (err error)
```
Done.

freeekanayaka commented on client/interfaces.go:
```diff
@@ -164,9 +168,20 @@ type ContainerServer interface {
+	GetCluster(password string) (cluster *api.Cluster, err error)
+	BootstrapCluster(name string) (op *Operation, err error)
+	AcceptNode(targetPassword, name, address string, schema, api int, pools []api.StoragePool, networks []api.Network) (info *api.ClusterNodeAccepted, err error)
+	JoinCluster(targetAddress, targetPassword, targetCert, name string) (op *Operation, err error)
```
I didn't look at this to deep. What you propose might work, except that there still needs to be an "Accept" and point which is different from "CreateClusterMember". That's mainly for internal use, because you need the joining node to hit an api on the target node to do some pre-work and return some info (such as if the joining node should be a database node). If we don't want to expose it we could probably remove AcceptNode and use RawQuery or something in the code.

freeekanayaka replied:
Actually, re-reading this now, the "Accept" endpoint would essentially be "CreateClusterMember" in your sketch.

At the moment BootstrapCluster() is pretty simple, it just performs a single POST /1.0/cluster/members request and it basically switches off the in-memory raft cluster and turns on the network version. We can easily rename it to UpdateCluster and change the endpoint to PUT /1.0/cluster.

On the other hand, JoinCluster is a bit more complex. Essentially it performs a request to the POST /1.0/cluster/members endpoint of the node that wants to join the cluster. Internally, the LXD API handler for that endpoint calls AcceptNode() against a client pointing to the target node of the cluster to be joined. AcceptNode() performs a request to POST /1.0/cluster/members, providing some details about the joining node (such as its networks and storage pools), and getting back some info from the target node (such as the private cert key, the existing database nodes, whether the joining node should promote itself to database node, etc).

If I understand you correctly, you're proposal would be to:

1. merge BootstrapCluster() and JoinCluster() into a single UpdateCluster() (changing the endpoint to PUT /1.0/cluster)

2. rename AcceptNode() to CreateClusterMember()

I believe 1. would be fine and not too difficult, as long as it's just a renaming/remix and we keep everything semantically unchanged, effectively (in particular the UpdateCluster API handler would still internally make that "accept" request described above).
 
One problem with 2. though is that this accept endpoint (however we call it) does not really create a cluster member per se, and probably should be considered somehow an internal endpoint.

So perhaps we could just drop AcceptNode()/CreateClusterMember() from the public API, leaving just GetCluster() and UpdateCluster(), moving the relevant logic somewhere else, and perhaps making use of RawOperation/RawQuery to hit the desired accept endpoint.

Hope it makes sense and it's clear.

freeekanayaka replied:
The new API layout and semantics are now in place.

freeekanayaka commented on client/lxd_containers.go:
```diff
@@ -71,7 +71,11 @@ func (r *ProtocolLXD) CreateContainer(container api.ContainersPost) (*Operation,
+	path := "/containers"
+	if r.targetNode != "" {
+		path += fmt.Sprintf("?targetNode=%s", r.targetNode)
+	}
```
Yeah I considered that, but for now I wanted to be really clear about what endpoints support a targetNode param, and not append it blindly. Ideally we should have some more declarative machinery that lets you tag an API as supporting that and adding the parameter if so. Considering that we already have a lot of duplicated boilerplate in the code, I'd suggest to leave this as it is, and perhaps tackle it as part of a broader normalization effort.

stgraber replied:
Can we do the enforcement on the server side instead, so effectively add a variable to Command on the server to indicate whether target is supported for that API endpoint, then if we get a request to any endpoint that doesn't support it, we immediately return an error instead of calling the target function.

That would let us always set target in the client code whenever UseTarget() has been used and force whoever uses the Go client to only use UseTarget() for actions that do support targeting, keeping everything clean.

freeekanayaka replied:
It still seems sub-optimal to me to blindly append a query parameter that is not going to be used (and add ad-hoc server checks that that you're not passing it when you're not supposed to). I'd rather leave this code like this for now.

freeekanayaka commented on client/lxd_server.go:
```diff
@@ -52,6 +57,12 @@ func (r *ProtocolLXD) UpdateServer(server api.ServerPut, ETag string) error {
+	// usually nil.
+	if r.httpUserAgent == "lxd-cluster-notifier" {
+		return true
+	}
```
Yes, it's a hack. As said, I think solving this properly requires a bit more change across the board, that's why I went for this. In some cases (especially cluster/node lifecycle events such as shutting down, joining etc) there's a need to not have the DB involved.

stgraber replied:
Let's replace that one check with:
```go
if r.server == nil {
    return true
}
```

stgraber replied:
done

freeekanayaka commented on client/lxd_server.go:
```diff
@@ -77,3 +93,26 @@ func (r *ProtocolLXD) GetServerResources() (*api.Resources, error) {
+// node-specific operations such as creating containers, modifying storage
+// configuration etc.
+func (r *ProtocolLXD) ClusterTargetNode(name string) ContainerServer {
+	return &ProtocolLXD{
```
Yeah I tried that at first, but static analysis fails:

client/lxd_server.go:101: assignment copies lock value to foo: lxd.ProtocolLXD contains sync.Mutex

We should probably add a clone() method.

stgraber replied:
Was that using shared.DeepCopy?

freeekanayaka replied:
No, it was using ```s := *r```. I just tried shared.DeepCopy, but it's not working since it only copies exported fields.

freeekanayaka commented:
Okay, I fixed the low-hanging fruits, in a nutshell:

  - Rename "/nodes" to "/members" in the API [DONE]
  - Rename a bunch of the client functions to match with normal naming [(PARTLY?) DONE]
  - Rework "GET /1.0/cluster" to show cluster status and system-specific keys (not the entire network/storage pool). [NOT DONE, see comment]
  - Use "PUT /1.0/cluster" as the way to enable clustering, supporting both the case of the initial node and of a new node joining. [NOT DONE, need some more work, and double check it's all sound]
  - "GET /1.0/cluster/members" should be consistent with other collection endpoints and return a list of URLs unless called with ?recursion=1 [DONE]

 - What's with that .gitmodules file I'm seeing on Github [DONE, removed]
 - Lots of mixed tabs/spaces in documentation [DONE, moved to spaces]
 - In API docs, we only list the content of "metadata" in output, it's expected that this is then wrapped by our standard response message (and described at the beginning of the API doc) [NOT DONE, it seems several other APIs as well list standard response message fields? should we take a pass and remove all them?]

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

stgraber commented on client/interfaces.go:
```diff
@@ -164,9 +168,20 @@ type ContainerServer interface {
+	AcceptNode(targetPassword, name, address string, schema, api int, pools []api.StoragePool, networks []api.Network) (info *api.ClusterNodeAccepted, err error)
+	JoinCluster(targetAddress, targetPassword, targetCert, name string) (op *Operation, err error)
+	LeaveCluster(name string, force bool) (err error)
+	GetNodes() (nodes []api.Node, err error)
```
Logic was reversed somehow, fixed.

freeekanayaka commented:
jenkins: test this please

stgraber commented:
So locally when building without `-tags libsqlite3` so just using dqlite with its builtin sqlite3, the tests don't seem to pass for me. Any chance you can look into this as it's going to be a rather serious problem for our contributors (none of which are likely to have a built copy of the patched libsqlite3).

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
@stgraber what failure did you get without -tags libsqlite3? I built lxd with a plain "go build", ran ./main.sh, and tests passed here.

stgraber commented:
@freeekanayaka looked like it was stuck waiting in leader election and running "go test -v ./..." would fail with some weird errors.

freeekanayaka commented:
@stgraber luckily I think I figured and fixed the cgo bug. Or at least it didn't show up in quite a few builds now.

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,40 @@
+// Cluster represents high-level information about a LXD cluster.
+type Cluster struct {
+	Name string `json:"name" yaml:"name"`
+}
```
I'd have expected this to look like:
```
type Cluster struct {
    // True if clustering is enabled
    Enabled bool `json:"enabled" yaml:"enabled"`

    // Same as Name but less likely to be confused for the cluster name (which we don't have)
    ServerName string `json:"server_name" yaml:"server_name"`

    // Global cluster status (rather than member specific), should be "Running" or "Broken" pretty much, "Inactive" or whatever our closest status is when disabled.
    Status string `json:"status" yaml:"status"`
}

freeekanayaka replied:
I can surely rename "Name" to "ServerName".

Re "Enabled", I mentioned that in comment above: do we really want it? As it's essentially redundant with ServerName. Currently, if ServerName != "", then Enabled == true, otherwise Enabled == false. I don't see a case were this correlation would not apply.

Re "Status", I can add it, but I'm not entirely sure what the semantics of it would be yet, that's why I'd suggest we hold it back for now and add it later.

freeekanayaka replied:
I added the Enabled field, as requested on IRC.

stgraber commented on shared/api/cluster.go:
```diff
@@ -0,0 +1,40 @@
+	TargetCert    string        `json:"target_cert" yaml:"target_cert"`
+	TargetCA      []byte        `json:"target_ca" yaml:"target_ca"`
+	StoragePools  []StoragePool `json:"storage_pools" yaml:"storage_pools"`
+	Networks      []Network     `json:"networks" yaml:"networks"`
```
This struct is confusing to me because I thought a whole bunch of those things were now moved to the internal call.
That is, the API client doesn't need to tell LXD about the storage pools or networks that it's got, nor its Schema version or API version. LXD itself already knows all of those things.

I'd have expected it to look like:

```go
type ClusterPut struct {
    ServerName string `json:"server_name" yaml:"server_name"`
    ClusterAddress string `json:"cluster_address" yaml:"cluster_address"`
    ClusterCertificate string `json:"cluster_certificate" yaml:"cluster_certificate"`
}
```

 - Address is core.https_address (PUT should fail if it's not set to something appropriate)
 - Schema is known internally
 - API is known internally
 - TargetCA isn't needed when we already pass the exact Certificate anyways
 - StoragePools is known internally
 - Networks is known internally

freeekanayaka replied:
Oh yeah, those fields were indeed leftovers, I should have moved them to the internal API. That's done now.

In addition to what you asked, I also renamed ClusterMember.Name to ClusterMember.ServerName, which seems more consistent. Those are individual commits, so feel free to strip them out if you wish.

stgraber commented:
jenkins: test this please
